src/huggingface_hub/inference/_mcp/mcp_client.py

```diff
@@ -0,0 +1,236 @@
+import json
+import logging
+import warnings
+from contextlib import AsyncExitStack
+from pathlib import Path
+from typing import TYPE_CHECKING, AsyncIterable, Dict, List, Optional, Union
+
+from typing_extensions import TypeAlias
+
+from ...utils._runtime import get_hf_hub_version
+from .._generated._async_client import AsyncInferenceClient
+from .._generated.types import (
+    ChatCompletionInputMessage,
+    ChatCompletionInputTool,
+    ChatCompletionStreamOutput,
+    ChatCompletionStreamOutputDeltaToolCall,
+)
+from .._providers import PROVIDER_OR_POLICY_T
+from .utils import format_result
+
+
+if TYPE_CHECKING:
+    from mcp import ClientSession
+
+logger = logging.getLogger(__name__)
+
+# Type alias for tool names
+ToolName: TypeAlias = str
+
+
+class MCPClient:
+    def __init__(
+        self,
+        *,
+        model: str,
+        provider: Optional[PROVIDER_OR_POLICY_T] = None,
+        api_key: Optional[str] = None,
+    ):
+        warnings.warn(
+            "'MCPClient' is experimental and might be subject to breaking changes in the future without prior notice.",
+            UserWarning,
+        )
+
+        # Initialize MCP sessions as a dictionary of ClientSession objects
+        self.sessions: Dict[ToolName, "ClientSession"] = {}
+        self.exit_stack = AsyncExitStack()
+        self.available_tools: List[ChatCompletionInputTool] = []
+
+        # Initialize the AsyncInferenceClient
+        self.client = AsyncInferenceClient(model=model, provider=provider, api_key=api_key)
+
+    async def __aenter__(self):
+        """Enter the context manager"""
+        await self.client.__aenter__()
```

Comment:
What is the point of entering the client context manager here?

It is only used within the `process_single_turn_with_tools` method, and the context manager is properly entered and exit there using a `with` block?
```suggestion
```

Reply:
Good catch! I initially intended to delete the `async with self.client:` below but forgot about it. The issue with including `async with self.client:` within `process_single_turn_with_tools` is that concurrent usage of `process_single_turn_with_tools` may result in unexpected behaviors if one instance terminates the sessions for all others. By relocating the logic to `__aenter__`, the responsibility shifts to the end user, who can handle the client lifecycle themselves.

Reply:
updated in https://github.com/huggingface/huggingface_hub/pull/2986/commits/b273cba153b8febfd01e5ef590459079ef34b26b